### PR TITLE
bpo-43913: Fix bugs in cleaning up classes and modules in unittest.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Functions registered with addModuleCleanup() were not called unless
  the user defines tearDownModule() in their test module.
* Functions registered with addClassCleanup() were not called if
  tearDownClass is set to None.
* Buffering in TestResult did not work with functions registered
  with addClassCleanup() and addModuleCleanup().
* Errors in functions registered with addClassCleanup() and
  addModuleCleanup() were not handled correctly in buffered and
  debug modes.
* Errors in setUpModule() and functions registered with
  addModuleCleanup() were reported in wrong order.
* And several lesser bugs.

https://bugs.python.org/issue43913

### CODE REVIEW

File: Lib/unittest/suite.py [hunk 1]
```diff
@@ -150,34 +150,38 @@ def _handleClassSetUp(self, test, result):
...
+                        raise
+                    try:
+                        failed = currentClass._classSetupFailed = True
+                    except TypeError:
+                        pass
+                    className = util.strclass(currentClass)
+                    self._createClassOrModuleLevelException(result, e,
+                                                            'setUpClass',
+                                                            className)
+                if failed and doClassCleanups is not None:
```
Comment (ambv) [inline Lib/unittest/suite.py:179]: `failed` will raise `UnboundLocalError` if `currentClass` is a function (builtin-type). We need to give it a default.

Comment (serhiy-storchaka) [reply Lib/unittest/suite.py:179]: It is set at line 153. I tested this case.

Comment (ambv) [reply Lib/unittest/suite.py:179]: No, you also need to set it in line 157.

Comment (serhiy-storchaka) [reply Lib/unittest/suite.py:179]: In `failed = currentClass._classSetupFailed = False` first `failed` is set, then `currentClass._classSetupFailed` is tried to set.

I rewritten the code in more clear form.

**Comment by ambv** [comment]
@serhiy-storchaka I'm fixing the conflict for 3.9.

**Pull request merged**